### PR TITLE
A hacky way to register service events handler dynamically

### PR DESCRIPTION
**Pull request by Huang-Wei**
#### What type of PR is this?

/kind feature
/sig scheduling

#### What this PR does / why we need it:

A hacky way with (_almost_) least code of changes to register service events handler dynamically.

- commit1: error out if ServiceAffinity is enabled but with empty args.affinityLabels.
- commit2: hacky logic. add a no-op indexer during New(), and verify if this no-op indexer's present during registering event handlers.

(alternatively, we can inject some logic to `frameworkCapturer` to be called back when a profile is composed - i.e., iterate all filter plugins, and check if ServiceAffinity exists)

#### Which issue(s) this PR fixes:

Follow-up of https://github.com/kubernetes/kubernetes/issues/94009#issuecomment-797716300

#### Special notes for your reviewer:

I've tested locally this PR works. Will add comments if this way looks good.

#### Does this PR introduce a user-facing change?

```release-note
scheduler: suppress registering service events handler if ServiceAffinity plugin is not enabled.
```

### CODE REVIEW

**Comment by alculquicondor** [comment]
Are you suggesting that this would be the way to support custom plugins?

If not, this is what I would do to keep supporting ServiceAffinity until we can delete it (I don't mind having the informers that much):
- Remove service events from `allClusterEvents`
- Have ServiceAffinity implement `EventsToRegister`

If you are looking for ways to supporting custom plugins, I think we should somehow expose `MoveAllToActiveOrBackoffQueue`. Then plugins create the event handler in their own package.

**Comment by Huang-Wei** [comment]
> Are you suggesting that this would be the way to support custom plugins?

I'd say I'd like to exercise the idea behind this PR, but the implementation is quick hacky yet. So any thoughts are welcome.

The idea behind this PR was to find a way to tell whether a particular API resource's internal storage/index has been instantiated (via the API's informer's `Informer()` or `Lister()` methods)

- if it's instantiated by at least one plugin, we should register the event handlers through the informerFactory
- if not, then bypass

However unfortunately, there is currently no **official** support in client-go - Lister() calls Informer() under the hood, and Informer() calls [InformerFor()](https://github.com/kubernetes/kubernetes/blob/0ed8ee6fd7169219fbeb9bca83023a980a7a36a5/staging/src/k8s.io/client-go/informers/factory.go#L162-L183) to ensure the default informer to be (always) present.

https://github.com/kubernetes/kubernetes/blob/0ed8ee6fd7169219fbeb9bca83023a980a7a36a5/staging/src/k8s.io/client-go/informers/core/v1/service.go#L80-L90

So in other words, once `Informer()` or `Lister()` is called, the API resources's informers lengh is at least one - there is no way you can detect empty informers. That's why I came up with this PR to exercise in a hacky manner.

**Comment by Huang-Wei** [comment]
> If not, this is what I would do to keep supporting ServiceAffinity until we can delete it (I don't mind having the informers that much):
> 
> * Remove service events from `allClusterEvents`
> * Have ServiceAffinity implement `EventsToRegister`

This is more practical than this PR, in short-term to figure out the issue.

> If you are looking for ways to supporting custom plugins, I think we should somehow expose `MoveAllToActiveOrBackoffQueue` in the framework handler. Then plugins create the event handler in their own package.

This gives users way too much power IMO. I'm fine with exposing an interface to enqueue a limited number of pods, but not expose a knob to churn all unschedulable pods.

**Comment by alculquicondor** [comment]
> This gives users way too much power IMO. I'm fine with exposing an interface to enqueue a limited number of pods, but not expose a knob to churn all unschedulable pods.

It's not that much power anymore, as they have to specify the event and not all plugins will react to all events.

**Comment by ahg-g** [comment]
For service affinity, I suggest we simply pass a flag to `addAllEventHandlers` to indicate whether or not to register the event handler for Services. The flag is set by the caller if any profile configures the service affinity plugin.

We need a proper design for the generic solution, perhaps over a google doc.

**Comment by Huang-Wei** [comment]
Per discussion with @ahg-g offline, we don't think we are able to make it for 1.21, so I'm gonna close it. I will work on a general design to fit requirements of in-tree and out-of-tree plugins.
